---
title: "clusterProfiler"
author: "K Chu"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Libraries.

```{r libraries}
library(clusterProfiler)
library(org.Mm.eg.db)
library(DOSE)
library(enrichplot)
library(lattice)
library(dplyr)
library(DESeq2)
```

Set working directory.

```{r setwd}
folder <- "/Users/karen/mount/chuk/Fig3B_betabinomial_on_LSC_and_LSK_unique_targets/"
```


Import ZiNing's data.

#```{r zining}
setwd("/Users/karen/mount/chuk/HOMER_HOXA9_MYC_MYB_IKZF2/input/")

lsc.zining <- read.csv("lsc_with_distance.csv")
lsk.zining <- read.csv("lsk_with_distance.csv")
#```


Import data.

```{r import zining data}
setwd("/Users/karen/mount/chuk/HOMER_HOXA9_MYC_MYB_IKZF2/input/")

lsc <- read.csv("lsc_with_distance.csv")
lsk <- read.csv("lsk_with_distance.csv")

# lsc.filter <- lsc [ lsc$diff.frequency >= 0.1 & lsc$ADA.fpkm > 5 & lsc$DCD.fpkm > 5 & lsc$MIG.fpkm > 5, ]
# lsk.filter <- lsk [ lsk$diff.frequency >= 0.1 & lsk$ADA.fpkm > 5 & lsk$DCD.fpkm > 5 & lsk$MIG.fpkm > 5, ]

lsc.filter <- lsc [ lsc$diff.frequency >= 0.6 & lsc$ADA.fpkm > 5 & lsc$DCD.fpkm > 5 & lsc$MIG.fpkm > 5, ]
lsk.filter <- lsk [ lsk$diff.frequency >= 0.6 & lsk$ADA.fpkm > 5 & lsk$DCD.fpkm > 5 & lsk$MIG.fpkm > 5, ]
```

```{r import data}
setwd(folder)

# Import beta-binomial output LSC ADAR vs LSK ADAR.
# df <- read.csv("Mouse_LSC_LSK_ADA_DCD_MIG_snps_count_dedupped_KNOWN_SNPs_removed_LSC_LSK_combined_diff.freq_greaterthanequal_0.6_significance_fpkm.csv")
# 
# df.filter <- df [ df$p.adj < 0.05 & 
#                     df$LSC.ADAR.fpkm >= 5 & 
#                     df$LSK.ADAR.fpkm >= 5 &
#                     df$LSC.DCD.fpkm >= 5 &
#                     df$LSK.DCD.fpkm >= 5 &
#                     df$LSC.MIG.fpkm >= 5 &
#                     df$LSK.MIG.fpkm >= 5, ]

# Import gene expression data.
reads <- readRDS("Yuheng_files/rnaseq_read_count_entrez_id.rds")
reads.mat <- as.matrix(assay(reads))
```


Gene Ratio = the number of genes in your input list associated with the given GO term (e.g., 5) and then a / and the total number of input genes

LSC plots.

```{r input format}
setwd(folder)

background <- unique( as.character( rownames(reads.mat) ) ) # All data
targets <- unique( as.character( unique( c(lsc.filter$entrez.id) ) ) ) # Targets

# ncg <- enrichGO(targets,
#                   #organism = 'mmu',
#                   #keyType = "kegg",
#                 OrgDb = "org.Mm.eg.db",
#                 pAdjustMethod = "BH",
#                 universe = background )

lsc.enrichGO.results <- ncg$Description

png("lsc.targets.from.zining.data_diff.freq_0.6_or_greater_enrichGO.png", 1200, 800)
barplot(ncg, showCategory=73)
dev.off()

png("lsc.targets.from.zining.data_diff.freq_0.6_or_greater_enrichGO_dotplot.png", 1200, 800)
enrichplot::dotplot(ncg, x="Count", showCategory=73)
dev.off()

# edox <- DOSE::setReadable(ncg, 'org.Mm.eg.db', keyType='ENTREZID')
# png("lsc.lsk.shared.targets_enrichGO_cnetplot.png", 800, 800)
# cnetplot(edox, foldChange = targets.diff.freq, layout = "kk")
# dev.off()

# png("lsc.lsk.shared.targets_enrichGO_cnetplot_circular.png", 800, 800)
# cnetplot(edox, foldChange=targets.diff.freq, circular = TRUE, colorEdge = TRUE)
# dev.off()
#
# library("pathview") # KEGG pathways
# # Hematopoietic lineage
# mmu04640 <- pathview(gene.data  = targets.diff.freq,
#                      pathway.id = "mmu04640",
#                      species    = "mmu" )

# # RNA degradation
# mmu03018 <- pathview(gene.data  = targets.diff.freq,
#                      pathway.id = "mmu03018",
#                      species    = "mmu" )

# # Ubiquitin mediated proteolysis
# mmu04120 <- pathview(gene.data  = targets.diff.freq,
#                      pathway.id = "mmu04120",
#                      species    = "mmu" )

# # RNA transport
# mmu03013 <- pathview(gene.data  = targets.diff.freq,
#                      pathway.id = "mmu03013",
#                      species    = "mmu" )

# Fanconi anemia pathway (negative control)
# mmu03460 <- pathview(gene.data  = targets.diff.freq,
#                      pathway.id = "mmu03460",
#                      species    = "mmu" )
```

LSK plots.

```{r lsk}
setwd(folder)

background <- unique( as.character( rownames(reads.mat) ) ) # All data
targets <- unique( as.character( unique( c(lsk.filter$entrez.id) ) ) ) # Targets

ncg <- enrichGO(targets,
                  #organism = 'mmu',
                  #keyType = "kegg",
                OrgDb = "org.Mm.eg.db",
                pAdjustMethod = "BH",
                universe = background )

lsk.enrichGO.results <- ncg$Description

png("lsk.targets.from.zining.data_diff.freq_0.6_or_greater_enrichGO.png", 1200, 800)
barplot(ncg, showCategory=18)
dev.off()

png("lsk.targets.from.zining.data_diff.freq_0.6_or_greater_enrichGO_dotplot.png", 1200, 800)
enrichplot::dotplot(ncg, x="Count", showCategory=18)
dev.off()
```

Venn diagram of enrichrGO results.

```{r venn diagram}
setwd(folder)

library("VennDiagram")

vd <- venn.diagram(
  x = list(
    "LSK enrichR GO results" = lsk.enrichGO.results,
    "LSC enrichR GO results" = lsc.enrichGO.results
  ),
  filename = NULL, 
  main=paste0( length(unique(lsk.filter$entrez.id)), " LSK genes\n",
               length(unique(lsc.filter$entrez.id)), " LSC genes\n",
               length(unique(background)), " genes in genome (background)"),
  col = "transparent",
  fill = c("cornflowerblue", "green"),
  alpha = 0.50,
  scaled = TRUE,
  ext.text = TRUE,
  ext.line.lwd = 2,
  ext.dist = -0.15,
  ext.length = 0.9,
  ext.pos = -4,
  inverted = FALSE,
  cex = 2,
  fontfamily = "sans",
  fontface = "bold",
  cat.col = c("darkblue", "darkgreen"),
  cat.cex = 2,
  cat.fontfamily = "sans",
  rotation.degree = 0
)

library(grDevices)

png(file="LSC_LSK_enrichrGO_results.png", 800, 800 )
par(mar=c(5,5,5,5))
grid.draw(vd)
dev.off()
```
















