---
title: "clusterProfiler specific functions"
author: "K Chu"
date: "11/21/2019"
output: html_document
---

User defined dataset.
TERM2GENE is a data.frame with first column of term ID and second column of corresponding mapped gene

Find RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO in data_resources folder.

```{r user defined dataset}
# Read in the data
x <- scan("~/Downloads/RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO", what="", sep="\n")
# Separate elements by one or more whitepace
y <- strsplit(x, "[[:space:]]+")
# Extract the first vector element and set it as the list element name
names(y) <- sapply(y, `[[`, 1)
# Remove the first vector element from each list element
y <- lapply(y, `[`, -1)

# Assign name in list as Term ID and genes as genes.
dl <- data.frame(term.ID = rep(names(y), sapply(y, length)),
                 gene = unlist(y))
rownames(dl) <- c()
dl$term.ID <- as.character(dl$term.ID)
dl$gene <- as.character(dl$gene)

# Change gene symbol to mouse convention
firstup <- function(x) {
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  return(x)
}
dl$gene <- tolower(dl$gene)
dl$gene <- firstup(dl$gene)

# Convert gene symbol to entrez id
entrez.id <- mapIds(org.Mm.eg.db, dl$gene, 'ENTREZID', 'SYMBOL')
dl$entrez.id <- entrez.id
dl.no.na <- dl [ !is.na(dl$entrez.id),] # Remove <NA> values

# Prepare TERM2GENE format
dl.enrichr.input <- dl.no.na %>% select(c("term.ID", "entrez.id"))
colnames(dl.enrichr.input) <- c("ont", "gene")
dl.enrichr.input$gene <- as.character(dl.enrichr.input$gene)

dl.enrichr.input.filter <- dl.enrichr.input [ 
  grepl(c("HSC|LSC|MEP|MSI2|MPP|Macrophage|proB|T cells|B cells|lymphocytes|Treg"), dl.enrichr.input$ont), ]
```

Run EnrichR

# Vignette example
gmtfile <- system.file("extdata", "c5.cc.v5.0.entrez.gmt", package="clusterProfiler")
c5 <- read.gmt(gmtfile)
egmt <- enricher(gene, TERM2GENE=c5)

```{r enrichr run}
setwd(folder)
# Run Enrichr on RNA-seq and gene signature
# targets <- dl.enrichr.input [ grepl("MSI2", dl.enrichr.input$ont), ]$gene # positive control for code
background <- unique( as.character( rownames(reads.mat) ) ) # All data
targets <- unique( c (as.character(lt.yuheng.filter$entrez.id),
                      as.character(st.yuheng.filter$entrez.id) ))
egmt <- enricher( targets, 
                  universe = background,
                  pAdjustMethod = "BH",
                  TERM2GENE=dl.enrichr.input.filter)

png("enrichr_output/RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO_plot_LT_ST.png", 1200, 800)
enrichplot::dotplot(egmt, x="Count", showCategory=length(egmt$Description))
dev.off()

egmt.hsc <- as.data.frame(egmt)
```




