---
title: "Gene expression analysis"
author: "Karen Chu"
date: "7/31/2019"
output: html_document
---

Gene expression analysis.
VST normalized counts will change depending on how many genes are in the dataframe.

Cells taken out from the same mouse.
Add dmso or tamoxefin in vitro and then obtain input or polysome.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Libraries.

```{r libraries}
library(DESeq2)
library(data.table)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(org.Mm.eg.db)
library(org.Hs.eg.db)
library(biomaRt)
library(ggplot2)
library("ggrepel")
library(data.table)
library(dplyr)
```

Working directory.

```{r folder}
folder <- "~/mount/chuk/METTL3_Hanzhi/DESeq2_aligned_to_reference_not_custom/KC_vs_WC/"
output.dir <-"~/mount/chuk/METTL3_Hanzhi/DESeq2_aligned_to_reference_not_custom/KC_vs_WC/"
```

Import data. Need to be RDS object in order for downstream stuff to work.

```{r data}
reads <- readRDS("~/mount/chuk/METTL3_Hanzhi/Project_10991_B_WT_KO_LSK_SON_OE/JAX_0453/align_to_normal_reference_genome/rnaseq_read_count_entrez_id.rds")

reads <- reads[ ,grepl("KC|WC", colnames(reads)) ]
```


```{r filter}
group.one <- "WC" # bottom of the log2FC(group.two/group.one)
group.two <- "KC"
comparison.name <- paste0(group.two, "_vs_", group.one)
reads.subset <- reads [ ,grepl(paste0(group.one, "|", group.two), colnames(reads))]
reads.subset <- as.data.frame( assay(reads.subset) )
```

Filter by fpkm >=1 

```{r fpkm}
fpkm.mat <- fpkm( DESeqDataSet( reads, ~1 ) )
fpkm.greater.equal.to.one <- which( rowMeans(fpkm.mat) >= 1 )
reads.filtered <- reads.subset [ rownames(reads.subset) %in% names(fpkm.greater.equal.to.one), ]
```

Prepare data for DESeq2. 

```{r prep DESeq2}
setwd(output.dir)

sample.id <- colnames(reads.filtered)
sample.id <- strsplit(sample.id, "_IGO")
sample.id <- sapply(sample.id, "[[", 1)

condition <- ifelse(grepl(group.one, colnames(reads.filtered)), "Group1", "Group2")

coldata <- data.frame(Sample = as.factor(sample.id), 
                      condition = as.factor(condition) )

dds <- DESeqDataSetFromMatrix(
  countData = as.matrix(reads.filtered), 
  colData = coldata, 
  design = ~condition) 

dds$condition <- relevel(dds$condition, "Group1") # intercept that DESeq2 calculates depends on the first factor. If "LSK" first, then results in log2foldchange(LSC/LSK). If "LSC" first, then results is log2foldchange(LSK/LSC).

write.csv(coldata, paste0("coldata", comparison.name, ".csv"), row.names = FALSE)
write.csv(as.character(design(dds)), paste0("design_", comparison.name, "_dds.csv"), row.names = FALSE)
```

Run DESeq2

```{r DESeq2}
dds <- DESeq(dds)
dds.vst <- varianceStabilizingTransformation(dds)
res <- results(dds, name = "condition_Group2_vs_Group1")
```

Choose below: org.Mm.eg.db for mouse annotation, or org.Hs.eg.db for human.

```{r gene symbol}
setwd(output.dir)

library(annotate)
entrezid.to.genesymbol <- function(res) {
  
  res$entrez.id <- rownames(res)
  
  read.count.entrez.id <- as.character(rownames(res))
  read.count.gene.symbol <- lookUp(read.count.entrez.id, 'org.Mm.eg.db', 'SYMBOL')
  # read.count.gene.symbol <- lookUp(read.count.entrez.id, 'org.Hs.eg.db', 'SYMBOL')
  read.count.gene.symbol <- unlist(read.count.gene.symbol)
  res$gene <- read.count.gene.symbol
  
  return(res)
  
}

dds.vst.with.gene.symbols <- entrezid.to.genesymbol(as.data.frame(assay(dds.vst)))
write.csv(dds.vst.with.gene.symbols, paste0("VST_normalized_read_counts_", comparison.name, ".csv"), row.names = FALSE)

reads.filtered.with.genesymbols <- entrezid.to.genesymbol(res)
write.csv(as.data.frame(reads.filtered.with.genesymbols), 
          paste0(comparison.name, "_DESeq2Results.csv"),
          row.names = FALSE)
```

PCA.

```{r PCA}
setwd(output.dir)

png( paste0( comparison.name, "_PCA.png"), 1000, 800 )
pcaData <- plotPCA(dds.vst, intgroup=c("Sample", "condition"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
#ggplot(pcaData, aes(PC1, PC2, color=Sample, shape=cell)) +
ggplot(pcaData, aes(PC1, PC2, color=condition)) +
  geom_point(size=5) +
  geom_label(
    label=pcaData$Sample, 
    size = 3,
    nudge_x = 2, nudge_y = 0 ) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  ggtitle(comparison.name) +
  theme(plot.title = element_text(size=40)) +
  theme(axis.text=element_text(size=30, color="black"), axis.title=element_text(size=30, color="black")) +
  theme(axis.line = element_line(colour = "black", size=1))
dev.off()
```

MA plot.

```{r MA plot}
setwd(output.dir)

png( paste0( comparison.name, "_MA.png" ), 1000, 800)
plotMA(res, ylim=c(-5,5))
dev.off()
```

Volcano plot. 

```{r volcano}
setwd(output.dir)

num.name <- group.two
denom.name <- group.one

padj.thres <- 0.1

res <- reads.filtered.with.genesymbols 

volcano.input <- as.data.frame( res[ order( res$padj, decreasing = F ), ] )
volcano.input <- mutate(volcano.input, sig=ifelse(volcano.input$padj < padj.thres, "Sig", "Not Sig"))

png( paste0(comparison.name, "_volcano.png"), 1000, 800 )
ggplot(data=volcano.input, aes(x=log2FoldChange, y=-log10(pvalue), colour= sig)) + 
  geom_point(alpha=1, size=4) +
  theme_minimal() +
  theme(legend.position="none") +
  geom_text_repel(data=volcano.input[1:50,], aes(label=gene), size = 5,box.padding = unit(0.5, "lines"), point.padding = unit(0.5, "lines"), color="black") +
  xlab(paste0("\nlog2(", num.name, "/", denom.name, ")")) + ylab("-log10(p)\n") +
  ggtitle(paste0(comparison.name, ";\npadj < ", padj.thres, ";\n# of sig genes: ", nrow( subset(volcano.input, sig=="Sig") ), "\n")) +
  scale_color_manual( values = c( "Sig"='red3', "Not Sig" ='black' ) ) + 
  theme(plot.title = element_text(size=40)) +
  theme(axis.text=element_text(size=30, color="black"), axis.title=element_text(size=30, color="black")) +
  theme(axis.line = element_line(colour = "black", size=1))
dev.off()
```



















